### DISCUSSION TITLE
Not clear how to use state API with an object

### DISCUSSION THREAD

**Discussion by brandonros** [vanjs-org/van#17] [Q&A] [answered]
Sorry if I missed it in docs.

```html
<!doctype html>
<html>
<head>
</head>
<body>
<script type="module">
import * as van from 'https://vanjs.org/code/van-0.11.0.js'

const { button, span } = van.tags

class Counter {
  constructor() {
    this.state = van.state({
      amount: 0
    })
  }

  onIncrementButtonClickFactory(self) {
    return (event) => {
      console.log(event)
      self.state.amount += 1
    }
  }

  onDecrementButtonClickFactory(self) {
    return (event) => {
      console.log(event)
      self.state.amount -= 1
    }
  }

  render() {
    const incrementButtonOnClick = this.onIncrementButtonClickFactory(this)
    const derementButtonOnClick = this.onDecrementButtonClickFactory(this)
    const incrementButton = button({ onclick: incrementButtonOnClick }, "👍")
    const decrementButton = button({ onclick: derementButtonOnClick }, "👎")
    return span(`❤️ ${this.state.amount} `, incrementButton, decrementButton)
  }
}

window.addEventListener('load', function () {
  const node = document.body
  const counter = new Counter()
  van.add(node, counter.render())
})
</script>
</body>
</html>

```

I was trying to teach my friend who doesn't know programming how to use a really cool lightweight library like this while showing him basic concepts of HTML/JavaScript/ES6/frontend development. I couldn't discern from the docs how to make this work. `this.state.amount` is `undefined`(`this.state.val.amount` isn't but that feels wrong) and none of the events work.

**Comment by artydev** [comment]
Hey Brandonros,

Look at this discussion [VanJSIntro](https://dev.to/artydev/our-first-hello-world-with-vanjs-the-new-mithril-4l1)
Read the message from Tao

Best Regards

**Reply by brandonros** [reply]
In the intro `const points = van.state(0)`, there is no object. I'm trying to make a more "real world" example other than state just being a simple number.

I couldn't figure out from the docs how to use this framework API wise when it comes to state getting/setting.

**Reply by artydev** [reply]
Look at this [MC](https://flems.io/#0=N4IgtglgJlA2CmIBcAWATAOgBwDYA0IAZhAgM7IDaoAdgIZiJIgYAWALmLCAQMYD21NvEHIQAHigQAbgAIesWqVIBeADog2ENgnUA+VdRkyAarWoApAMoy2LeDICMGNAGsAQjICqASRmEATvTwAO58LgZiAPSSUvrUETEy0GogtAAOaXoGRhHR0nEG3CCk8Ag8mgLkTAAMSA4A7CAAvng0QaIYAFbkvAJCIkwGEGBpfP5sMlJmfv58YDLq7GxppEiRkVPU3RhjAOaR-FDwG2YAtNUYDk4AnBiQ1F2k6gYG-NSkE8AARgCubGwCPAyGJAlgOJoyZSTMwYNi0XZPeLUN4fGSWOFCSHQh4fWhCAAUwGy0NgP3g1RkSBk1TwxKmpPgDkp1NphhJZLQzOqBiaAEoXsjKhNaOUIJUsfj8bihLzIboZES2SiJqQsdL4Bh6QBuYn+eBsH7+QyE4lGCDI5n481HAAesuU8sVRmdMnVmtosCxwAwPtIQIoAAN6WSACTAa3wG1NAMAXWZAGp46RA8H4GGI1HY01TTIWjmjjxLRn7Y6c0Y3fSvT6MH6ZCmPaHw9RbdG41TTqdk0GG2mmy2szmIc7iXyebz8ei8fB+UjCD9kRVDABhPjzoT+KUY+BAjNY6qyp1GPUGo3A6QyfE5zYYL7WzdToFSksyMGXkCmBnM9RArup9PNyNW15XlWRdX5-gEQkBHkCAeBcS1nxFRdSAwc0eCtAC7RaBYQG8AA5Jd1BAnNwIBagoORWBYPgql8UQ0VKgwAsMNtXlsPUAARABRQiQFlYkZ2zWd5wYwwAEEMgvdUD11fVDUMGJLzZIwKGrMT-ECABPfEAGZeRjO50klAB9HdMOfFc13gDd1TM1iZyMQSBWvWgYHxdI0iBCS0gnLdgIFQoCBKMpkNEWpqlOFBwuaVoQDoBgOh4JQijefo2FEL4+CgTSFWJL4RRcXZZnnKAqS+BQ4J1JU+FgMYqSgWh-BcMYzF2eAqqMQg+lOQh6BITSqUaiAPSqoSDFhLQEFytkhBtNhTg9CBdmoJAeGEdcOr8brSAgAAveAqTQPUwE2tJXMkahdipHSAFY0htUaBQAYg86ajHyuCitXZsqX8XZ8vxNAbpuvAHBu-AaQuBxeVO87zSu6lHGqe6ZDQZGbSR+7NuCaBbEOm70c22b5sW5bVvW6zNsy-wjn8KkHBR0gaugGQWsu9riTARrdnNKkKVoP4+EepFSIEN6ZGp2mkGoAQObZD7CuKn62ba2GYHhqk7oe4kusETs9oOsGsZ17rgngJb2CQTLYCgTatr13rIFgAaZCGkbTUl6z6cZ5moAliqXGFgw0j1cXdfmnb9vprWeSKYL4FEqoQAca46maGMCHixgQAAWWMTxTAsSxmiAA)

**Reply by artydev** [reply]
Beware not to apply your existing knowledge with VanJS.
You perhaps don't need Class for most projects...

[Forms](https://flems.io/#0=N4IgtglgJlA2CmIBcBmALAOgOwE4A0IAZhAgM7IDaoAdgIZiJIgYAWALmLCAQMYD21NvEHIQAHWoSAPFAgA3AATQAvGJC0ADhrUA+aQHpZcvZOrcQpeAh5sIA8kwBMSAIxoQAXzw16jEHNpqDAArcl4BIREmBDYFeDw2PGo8WGUAeQAjYPgbPAAndKyctgwAc3g2AAU8vjZagE8NeDTCPFJlAAp4xOTYAEplHS6AfmGOyzYAFQgGPgBXNg7evqT4AHcFAGUKvr6MWhgOthWoZWBy2IDYDr7gPIq5vOoFNhYIUgw2LwmFK67bmIKajKV7vFLKIJsADc8AAhMplLAAGRIpYYVII2DDNjKUhHJJ4Hh9JDwBEQ9EokpQKwVeBLFaQ5TxIJ8DCEPh5ACitB4LA6R0Gi3i-V2fS8Amo63+wFBH1ZGjmpD58DF4uUeQ6UBWECZg3gGGofGpk0a8GG8CQko2k3gAA8hSs5p14hhXcdBkc2bBaIsXPoAAx7dlcnl8gU6fWaJrUKAAYTesCgHQ17oRUGGa3xXT1uyQECOopWKrwtAh6wU1T4tvq-PxroNAx0wBiFFIeHrEAAuuqlhR-Z3RoGyXxhtQkFQvPXqJ2S8ooHweHMGIIMDx7j74JyEMvFscofc2I9nrAMMI2HkIPA8aQgxzubzaxQep3G82KgpUrQn534UzzYMvzYbsLVJHRaAwCYAEE2HPCAMgWOlEhVfd6TJdNM2SbMhmuYtaFFJANWoAYERHTMpwwakNDbfl6xVD1riCQgOlo3YS3whixVYuYOlodtXQgTjvAuJB8SIwUMAyCAYy6JJBIAfV1CMMB1fUIDZEghA1LD9TmYZlNIWMJWKeAtRWHhOlfQFgQoesgKhHFmXLbZFkY702AAWU0Do5N2Nk71DfkcSbGJgDky08DmJBYDwUoCI8EE2h7ciwC87TPlY6LktS0D9RFPooVIX9kVRQq-xPe4NG9Hh4AAdQgV5PSdfMb2JDpyvgMA+DkOkVhKJ0VU4hk-JDB80tSfVjjFPA1k6WzLPfVISlIWAIGqjp-TwABaFwVgKNgn3RYRSlebaZ1ODV6xPFKNH5HL0t2KEICY04mT6A8jwoTsoUBPgzjC6LIvzLUYri-cHieD9hvvMMguAYE8WSRJ+UbNFg2h7yGVelYXHgFBeuUjAFSVDo+EGvBWTmDwYVtDQOVialCFoOZYDYYADigJA5gSWhSlIJBeNINgNxJD1QrkjQajqOSwqgBISRSeWIHHGc+GVjwVkkmMkDWKnzEsaxbHsUR-VcFwcBwTxvBAOgGFEEIwhAfhBDPUQZlpvJLkCBRCBqMAFDUDB9ACIJQjUCQJCdwWFGABRSA0QI8AUFgXEToxE+DMBE6khVEg-WgMisRP4JggRE5YPIFA8BRlF+QJPh50hw+oSPYnZ6uFA6DJaEsPpq50DurGOPva6CdnO+7+I4hZvom5bl5aDyC52-nRcdzKCotw6s8ACF6gASSTNQozUfKm4kQg5mbw3nmW+B4BusAb2jiQFAUd6IatCtffeOkL6vuxnjJl7sAF+r9Y4VGmLMBYyZE6Pxns8Su8CPBNz-jYABCgABiHIwA3Gfgg0Bc9QGvxtvAduwcIJCyEB0AA5NQlYRC4gpRIGQuugsNw0LoXgBh8A8g1ArjXchbCqG0PoQg1+hA75QC7jwAA1iwoIQi6QiKhAw0BgIM41TyFGHhKixH+wQXPFggQ4DwAAHK+HjIEco7cujdUEL3QYeCwFAl8BgK4Nj4B2JKELReFQ3G0FgHMZCoCqZN1foY4xCBORMNgJY6g1ia62LPA4-uIC9EdVoCQfxsAPFePrr4koVwgmnwQaE0w4T7CxCMTGBAmw5gZEgLERJnjknDzSc4lpK4JadLYAAEXgIzZmiwSnOIkSZaRMjsntzDiAXRzi54Zy3E0722DNHaLyBgAAjkEvI9RtgGw5B0NQGcT7OLCWAp6HdYQkKmQAH1uQoWEGSslXGAQw1+PC+FTMSdQyoCAJ7exIDkqSCgAk5OIFYKAHw6FzOcW-cG1BYWvyrqA7hvCOTfP9uIWZRCiG33vh0AArP6QMnwWDCGRsPMZUieSTPcTXNQWC8h+1IPUxpQgoAn10Qw-FN1HDEtJa8CluDHHR30XChQFASGJ2edFOI6K8idihgFQgVLMUzJGXC6lEz1XYqRWAxZLMMD3AmDcfV6tVGlO5QgjRWitA8PbgoIwRy9HJyOSAAAEpEqSpRMHYNIEoZ4AA1QIAApTYJ8uF6Izi6iVzqY48G9KQPmWKIWJjUInAQvIrEWiTpEsxFjqnWK8OKiVecC7XDUOYhgkb3kXOoDnG4oi43yA6AmpNKbjmXnTdwBQWai25uqSY6JmTYkDsronOtr9vQVvdZ5Egta9HOOztA1idb40KETd3TtFg2X1TUBO0tEri51GoG2qdYC2CmiQFi1lDT93cAva-LNK1ZE3qHbUvdbA60lrUHexpJ94HOObWA8uNwo3OI3Vu5NN61CwD4KUcoeQD1eHXa24A0G+ZwYQ2mzllssXVtIa-WDvaSEgcg+hzDSBsOlFw2oX9IAR3MIUCRmVMTyNgPjVRmjdH8NqE5AqljWKZUKo46-LjHbqMgHg7R7teGGMYMkV3ORrHvZKdpUBsBmnNPv2eLa9ZEhkGmFQdfV+TDAFvIQWPKMfQOhMpwbsQzTdzM3DPmYAg+tigAIcCAAAHAANiQP6TaJt-SWx8LbJgq5k3mCdpENgogMhGnqE4hQEzSg1EvhzNLVUZGwv4PBvIN6oALxkRyHNsL2SCE2ozSAsB6g3oXhAAJuijMSAAMRRjFaA9LmWtZv1KF3Do-LCV4BcIS-zeANr+gwI4HasL44wB9Texw-oNC2gUKt9bm21sba27aWFaxoCvBvYS4l63YVCHtJtAJEBShjmqs7PIsKUqLykje-0oKFh8Fa03dSkLUtHagCdzb52DugNe6Ud7ChPtMzqC9heUPqCbSSyXMArhdu-dMDOqwqXZBx29A1qSK1JQo-g7Iy7do2A3ZWvdgid32AI7e8ji8pR2AY4u6AoHIP-OY6c6YCCX7AfHZYEgRw+2mdI6C19+HoCrvU9u3T1n7B+cSEF-e2IJ6BCpaS3kakRXDSSlhb1+Y-XyvxPgFjtXMmkOpcW7IeJm06gaA5+DhB3PRcoBJZzhBkPoew++5T67iuxwIEIN+6gbWggydw6lv3yPUd1HRy4Pnke9Y0jQUbJgKBfNIAth4Gc1tfCiHckGgAqiG6g9nyAeCAA)

```js
import van from "./van.js"

const { span, h1, div, form, input, label, button, hr } = van.tags

const add = (base) => (elt) => van.add(base, elt)

const target = document.getElementById("app");


function sleep(ms) {
  return new Promise(function (r) {
    setTimeout(r, ms)
  })
}

function Form() {
  
  const 
    name = van.state(''),
    email = van.state(''),
    error = van.state(''),
    feedback = van.state('');
    
  let formWrapper;
    
  const handleNameChange = (event) => {
    name.val = (event.target.value);
  };

  const handleEmailChange = (event) => {
    email.val = (event.target.value);
  };

  const handleSubmit = (event) => {
    event.preventDefault();
    feedback.val = "";
    const formElt = formWrapper.querySelector("form")    

    if (!name.val || !email.val) {
      error.val = ('Please fill in all fields.');
      return;
    } 
  
    error.val = "";
   
   sleep(500).then(() => feedback.val = "Form submitted");

    sleep(2500).then(() =>  { 
      [name, email, error].forEach(f => f.val = "");
      feedback.val = "";
      formElt.reset();
    })
    
  };

  formWrapper =  div(
    h1("Handling Forms in VanJS"),
    form(
      div({ class: "field", onchange: handleNameChange }, 
        label("Name"),
        input()),
      div({ class: "field", onchange: handleEmailChange }, 
        label("Mail"),
        input()),
      div({ class: "submit" }, 
        button({
          type: "submit",
          onclick: handleSubmit
      }, "submit"))
    ),
    hr(),
    div({ class: "logger" },
      div({class:"logfield"}, "Name   : ", name),
      div({class:"logfield"}, "Email  : ", email),
      div({class:"logfield"}, "Error : ", error),
      div({class:"logfield"}, "Feedbak : ", feedback)
    )
  )
  return formWrapper
}

function  main () {
  add(app)(Form())
}

main()

```

**Reply by artydev** [reply]
Another Example

[Demo](https://flems.io/#0=N4IgtglgJlA2CmIBcA2ADAOgKwHYA0IAZhAgM7IDaoAdgIZiJIgYAWALmLCAQMYD21NvEHIQAHigQAbgAIesWqVIBeADog2ENgnUA+VdRkyAarWoApAMoy2LeDICMGAEwBrAEIyAqgEkZhACd6eAB3PlcDMQB6SSl9akjYmWg1EFoAB3S9AyMJaWSoVM1teGzDGWjY+MrpeINuEFJ4BB5NAXImAGYkNBAAXzwaYNEpMwwAK3JeASERJgQ2GXg8NjxqPFhlAHkAI3H4VrwA7b2DtgwAc3g2AAUAvjYHgE90+C3CPBUACmXV9dgAJTKXQ-AD8oK+TTYABUIAw+ABXNhff4AtahGSWa4AgEYWgwL5sNGFYBXRajWBfAHAALXBEBQy2CCkDBsAZQmQUn7UhYyajKJmkDbKaisgDc8AAhMplLAAGRylEYTYy2CgtjKUiEtZ4HgApDwGUi5UK85QZrXeAotGijXLUV8DCEPgBACitB4LC+hOByOWgJxAIGAmooW5wEFGEd6QRpC98CDweUAS+UDREGUht08Aw1D45uhL3goPgSFDIRk0PgAA8-WiEcofngMC2icDCU6FMiHFE0LjnW6PV6fdm8ZlhFAAMIsEhQL4ptsyqCgkLan7AhP6iCEwNohN4WgijF3PjVp7e7Ut3NA3TABYUIVXiAAXWTKIoaGf4L7Rr4oOoSBUAMV7UM+B7KFAfA8AiDCCBgPC0rQQiuggsHIkSYq0mw9KGLAGDCGwAQQPAWqkP2Lrup6F4UH8z43ne1wyJstA0c+0qZiWwIsWwr6llmtAYFCACCbCERAOxIlaqwJph1pGsuq7rOuIKUvutCBkgKbUECMp-quIEYOa6RCt6V4Ju2lKioQXxmTiB4aZZQZ2QiXy0M2LYQE5gxkkg2rab6GA7BA1BzvaXkAPqZhuGAZjmEBOiQQgpspOYIqCMWkJOIZnPAaZojwjb0by-IUFePFina6IVliyJWV2ACyGRfOFOJOhRQ7ehqt4LMA4VlngCJILAeAXJpfQCp8b4GWATUpaydnDdNs1ZjmAYAmKpDsfKiqbRxeG0ukCg8PAADqWheucDbbmR+pfPt8BgHwUhWmil2ZjifR2VZ7VUXNmw5kSQZ4CEjZlUVjGbOcpCwBAx1fGgeAALQOGixxsDRyrCBctjI2BhQpleeEzek3orfNOJihA1mFO9WE4RQz5iryfDKL1Q0DUg25piNY2YXSDJMW1g6-V1wD8lq6yrN6N5KgOlFei1azvWiDjwJ0r0xRgMZxl8fBOWijoIn0ErVukLqLOahC0AisBsMA+JQEgCIrLQFykEgbmkGwSF8e2vXhek9yPOFfVQCsBobBHECAWBfAx59eBBSFSAhMbDRNC0bTUB0IBoEgDgOAAnIX-SDCAdAMKIExTCA-CCARohwmbATkv49xgDI6gYFEoyipM6gGAYddezIwCkOkZh4DILAOFPsRT3CFwyH0MjKJyrKu6Qg-UMPiwO6vMhfDsigJqvuiH80RJn+vDtHyfU+XwC2+7zYtABGSB+QdBaGXNcKEPQRdwTwfBznUBkLIIAn473aIsYoCBP5QRggRX+bB-5oSASAr46g4GlEgc-GBMgvZPAQOFWM8AAgB1hosNeAADHIMgdgunNAEfO6RqyEL4DDKAMgXRmCuGKehE8YDBVGjITobCBHlBmu-YK+c0BsJoZIgwhAEQ7yzt4JoARD5kICACUe9C6YCy+EYTu5QTEyCiFEDR5CZAhCCOOAI9DzFJC+GPNgxDSzqEkOPBQTwkCEAQNWMUMhaAwwuNQRGWgHru2OvXAIYoQjQFsEgQuaAACkAjS5OPMRYqxXhNEyHSFQ+kpQzE5NiK4ohCAkDqGkRcYKiMiIXHYHIth6gBg5OyTk5IYALiVPcdUqp8BSGaMoYcQhAQeBO00VrYptJWQsBgjsOgJBPpdJMWidZRhLHWK0RXUp3SjAuK2SY8eZgvg6NzMEBKAQvZT3UJ3SBU9DmmJeUYM51ALkzP2cqRQRITlQO6YCjZBg+jbxUWoiAAgbB8HMKQAQ849HAAMfzQwARq4IqgWChI1AIWtChYYbxh1aBPHyeQ0g2jNFkX0WY2+OCASpmkFgkAAAlMwkEO5ktuTIISNwfDqBxK8owOiWS0lIDbNgLJiZfJscCEJBJvbv2xFgspRguUyt0cCmQejTEr1Mco64VF1DsDYMZJAligghT4GAS5DAogZAgKCMVEqVBYAFfQhZwhCSwvhdpD1tgvVEt8Vysi9D6HpwtPi9oogAAsWAkDOERnnXoAwhiVyYPBJQDQ66zDYKIRhUAng0qMMfHgrgLj3FUY7BhR1XCSKMPwWALokAyCgG-VwvDqD8Poc6QQiMraQFgH4kJRFQmSOxQYVkWh4HIvKEIWsiNQkQHCUgWJSV63+BmIjUgEAABepYZDOFpGADdQjJBdpbZ0LAEjQXbwAMTgNHkK0t5bK3JxkO-Y+XxnBYCwHgBwWAUB4ARpgZwKNT0OxES25w8j2EwbYYe2DiGb3lESVAZJMhf2wY3fOtgi6wkATXeQjddTZEyDQCEpEfBx3bwjZnAlOcYM9H6GBcuwwmD1WMF4UwFhLAhv6EAA)

```js
import v from "./van.js"

const {span, h1, div, img } = v.tags

const add = (base) => (elt) => v.add(base, elt)

const target = document.getElementById("app")
const title = document.getElementById("title")

const style_user_pict = `
  border:1px solid orange;
  padding: 3px;
  margin:10px`;

function User (user) {
  return (   
      // User wrapper
      div ({style:"display:flex; align-items:center;width:90%;"},
        // User picture
        div({style:"margin-right:10px"},     
          img({style:style_user_pict, src:user.picture.thumbnail})
        ),
        // User name
        div (
          span(user.name.first, " "),         
          span(user.name.last)
        )
      )
    )
}

function toJson(r) {
  return r.json()
}

function displayUsers (users) {
   add(title)(div("Random Users API")) 
   users.results.map(user => add(target)(
      User(user))
   )  
}  

fetch("https://randomuser.me/api?results=5")
  .then(toJson)
  .then(displayUsers)
  
```
```

**Reply by Tao-VanJS** [reply]
Thanks @artydev!

I was about to write an answer to this question and saw you already have it answered pretty well. Here are things I would like to add:

For @brandonros's sample program, a more idiomatic way to define a root-level state is to have a JS object with VanJS states as its fields, rather than having a VanJS state with JS object as its value. Doing so can generally lead to performance improvement, making your app more "Reactive". It will also make the code to construct DOM tree more ergonomics. In other words, instead of having:

```javascript
this.state = van.state({
  amount: 0
})
```

It's more recommended to have:

```javascript
this.state = {
  amount: van.state(0)
}
```

Based on this style I have modified your program to make it work, see it here: https://jsfiddle.net/ptox6ws1/7/

btw: I saw you're using the VanJS version 0.11.0 in your program. The latest version of VanJS is 0.11.10 as of now. While theoretically 0.11.0 should work but this is a development version (created before we publicly released VanJS) thus we have relatively weaker guarantee on its reliability.

**Reply by artydev** [reply]
And alast 

Not the recommanded one :-)

[Demo](https://flems.io/#0=N4IgtglgJlA2CmIBcA2ALAOgIwHYA0IAZhAgM7IDaoAdgIZiJIgYAWALmLCAQMYD21NvEHIQIAL54a9RswBW5XgKEimEMAAc+AJzYACAG61qewtr5g9AcnZsNpJAHpHR6gow6A5o-5R4L4wBaAAZsLAwATgxIagwFKwAdaiT+alJ9YD0oCAM8PQAjAFc2NgE8iGoNYr1xPQBeQ2MMNlpPUgBuJJSBdL0AZRahesbY9NohAApgJL09LQr9JD0ADgi8Gb06BiWEkAAJWlhdpPEASk7k6gR9Wh42CB7ZhomJsaFT+oA+PWmTPVTekZYMM3vAMECNtp4GxCtoTFM9BtZhUeHolhMPnVvqDwYdhsAMISgXl5oI0XoANQUoEYUn6WobM4nU4TAbjeDnLrUQiFah3B4mACCGg0eleg3gQI+v1m1z0pBBEohfwB+gMEHgAHdhtkDGKkbNda8MFsOes-rMsjljXTTubLbMiiUBFMBDxYBAeABrcm3e49DAomp5XYo3anDYRv5QmFwwwazUXcRc1wYWgwCZQPg8QoMQQYfJ8KAATzyqfyFSgrIleWFGlOnMuXK53BApHgCH5PVEKCQWAAzBIpCBTaIMDxSIoQKkVGxRDx4IFNRA2CxApLhBkNvxYDollCoCc9K3253-WlRFg+wAmCQAXQIo6YADVjAApPoAYT4vKE2gAYuYYAAPL5HI8B3BIQA)

```js
import van from 'https://vanjs.org/code/van-0.11.9.min.js'

const { div, button, input } = van.tags;

const State = van.state({
  point : 89,
  name: "Hal"
});

let actions  = ((state) => {
  const val = state.val
  return ({ 
    inc : () => state.val = {...val, point : ++val.point } 
  })
})(State);

function App (stateval) {
  let s = stateval
  const view = div (
     div(s.name),
     div(s.point),
     button({onclick : actions.inc }, "inc")
  )
  return view;
}

van.add(document.body, van.bind(State, App));
```

**Reply by brandonros** [reply]
```html
<!doctype html>
<html>
<head>
</head>
<body>
<script type="module">
import van from 'https://vanjs.org/code/van-0.11.9.js'

const { button, span } = van.tags

class Counter {
  constructor() {
    this.state = van.state({
      amount: 0
    })
  }

  onIncrementButtonClickFactory(self) {
    return (event) => {
      console.log(event)
      const oldState = self.state.val
      const newAmount = oldState.amount + 1
      self.state.val = {
        ...oldState,
        ['amount']: newAmount
      }
    }
  }

  onDecrementButtonClickFactory(self) {
    return (event) => {
      console.log(event)
      const oldState = self.state.val
      const newAmount = oldState.amount - 1
      self.state.val = {
        ...oldState,
        ['amount']: newAmount
      }
    }
  }

  render() {
    const incrementButtonOnClick = this.onIncrementButtonClickFactory(this)
    const derementButtonOnClick = this.onDecrementButtonClickFactory(this)
    const incrementButton = button({ onclick: incrementButtonOnClick }, "👍")
    const decrementButton = button({ onclick: derementButtonOnClick }, "👎")
    return van.bind(this.state, (state) => {
      return span(`❤️ ${state.amount} `, incrementButton, decrementButton)
    })
  }
}

window.addEventListener('load', function () {
  const node = document.body
  const counter = new Counter()
  van.add(node, counter.render())
})
</script>
</body>
</html>
```

I was missing `val.bind`. Thank you very much!!

**Reply by artydev** [reply, marked as answer]
But try to follow Tao advise :-)

```js
const { button, span } = van.tags

class Counter {
  constructor() {
    this.state = {
      amount: van.state(0)
    }
  }

  onIncrementButtonClickFactory(self) {
    return (event) => {
      console.log(event)
      self.state.amount.val += 1
    }
  }

  onDecrementButtonClickFactory(self) {
    return (event) => {
      console.log(event)
      self.state.amount.val -= 1
    }
  }

  render() {
    const incrementButtonOnClick = this.onIncrementButtonClickFactory(this)
    const derementButtonOnClick = this.onDecrementButtonClickFactory(this)
    const incrementButton = button({ onclick: incrementButtonOnClick }, "👍")
    const decrementButton = button({ onclick: derementButtonOnClick }, "👎")
    return span("❤️ ", this.state.amount, " ", incrementButton, decrementButton)
  }
}

window.addEventListener('load', function () {
  const node = document.body
  const counter = new Counter()
  van.add(node, counter.render())
})
```
